lib/generateTasks.js

```diff
@@ -6,6 +6,22 @@ const path = require('path')
 
 const debug = require('debug')('lint-staged:gen-tasks')
 
+/**
+ * Match the following characters for escaping:
+ * @see https://regexr.com/4vt72
+ */
+const ESCAPE_CHARS = /[|\\{}()^$+*?"'` ]/g
+
+/**
+ * Match square brackets due to special case in Windows due to a bug in `minimatch` which `prettier` uses
+ * @see https://github.com/isaacs/minimatch/issues/50#issuecomment-258709394
```

Comment:
@iiroj After a lot of debugging, I came down to the fact that `prettier` wasn't handling the `[]` characters properly. https://github.com/isaacs/minimatch/issues/50#issue-40122775

Not sure we want this in the actual `lint-staged` repo since this is supposed to be more generic and not apply fixes for `prettier`

Reply:
I'm thinking about putting a PR in upstream for `prettier` regardless but I'm not sure what you want me to do here.

Reply:
Are there any other cases where escaping is necessary? If _lint-staged_ provided a generic safe escape option, so that projects wouldn't have to [configurations like this in Next.js](https://github.com/zeit/next.js/blob/canary/lint-staged.config.js), then I think it would be worth it to add in.

Reply:
Sorry, I'm not sure what you're asking.

I think that this `[[]` and `[]]` is a workaround for tools that use `minimatch` but may ultimately cause issues with tools that properly handle `[]`.

Reply:
If it’s strictly a minimatch workaround, I’d say it’s not _lint-staged_’s responsibility.

What I meant with my previous reply was that _lint-staged_ should try to handle all escaping in a generic way, if that’s even possible. I think wrapping paths in quotes would be the best way to do it, but unfortunately execa/spawn seems to strip those away.

Thanks for taking a stab at this!

Reply:
hmmmm... 🤔 I just linked my `prettier` PR with `micromatch` and it still didn't seem to fix the issue. Although, I'm not entirely sure that I properly linked it all up. 

>What I meant with my previous reply was that lint-staged should try to handle all escaping in a generic way, if that’s even possible. I think wrapping paths in quotes would be the best way to do it, but unfortunately execa/spawn seems to strip those away.

Honestly, I'm not really sure what the best route is for this. I'm far from an expert on the matter. Just trying to get it to work on Next.js without a huge workaround.

I will keep investigating this weekend

Reply:
Since lint-staged is handling globs, this will be an issue with any tool that does not have an option to treat input filenames as plain strings. 

It is not really a bug with minmatch, it is doing exactly what the docs say... it has no way to know this is a plain file name. 

The referenced bug is not really minmatch's fault, bash turns `\[` into `[`, unless you do `\\[` or "\[stuff\]"... but it happens before JS code even sees the strings. This is a pretty difficult problem for portable software, using a library is likely ideal.

Since there are several different JS based glob libraries, trying to cater to the unique flavors of each is going to be a mess.

Reply:
@sever1an Yeah, _lint-staged_ doesn't use a shell to parse anything by default since it directly spawns processes via execa. This way using quotes doesn't really help anything, since sending a `"/foo/[bar]/test.js"` argument to the task would mean the task itself had to strip quotes before parsing further.

Reply:
Escaping brackets with backslash and wrapping the name with quotes does fix the issue, it is what next.js does in their repo. I assumed it was a bash thing, but as you mention it is not possible.

Could the prettier glob library do something sane with big quotes wrapping the name?

Reply:
@iiroj After a bit more experimentation using `xarg -d ' '` to skip the shell, I was going to write up some findings, but honestly the whole thing is a mess. 

This works, but it is not immediately obvious why... the `"`s around filename are necessary. The `shell-quote` package escapes `[` and `]` with a single `\`. 
https://github.com/zeit/next.js/blob/canary/lint-staged.config.js

Honestly, I think using `\\[` and `\\]` work as a subtlety of the globs being turned to RegExp, and not due to any conscious design decision. 

I am voting for the issue at `prettier`. Sorry for making mis-informed noise.